src/features/devices/hooks/should-be-managed-by--release.ts

```diff
@@ -0,0 +1,34 @@
+import { hooks } from '@balena/pinejs';
+import { getAffectedIds } from '@balena/pinejs/out/sbvr-api/sbvr-utils';
+
+hooks.addPureHook('PATCH', 'resin', 'device', {
+	PRERUN: async (args) => {
+		const { request } = args;
+		const deviceIds = await getAffectedIds(args);
+
+		if (
+			request.values.should_be_managed_by__release != null ||
+			deviceIds.length === 0
+		) {
+			return;
+		}
+
+		await args.api.get({
+			resource: 'device',
+			options: {
+				$select: ['supervisor_version'],
+				$filter: {
+					id: {
+						$in: deviceIds,
+					},
+					should_be_managed_by__release: null,
+				},
+			},
+		});
+
+		/**
+		 * These device(s) should have a valid release assigned based on their current
+		 * running supervisor version, if we have one...
+		 */
```

Comment:
It looks like you're missing code to actually use the GET result?

Reply:
So, it seems to me that the intention of this code is, at the moment of PATCH, if `should_be_managed_by__release` has not been provided and the device needs `should_be_managed_by__release`, then patch it with the current supervisor version.  

If I am correct, then something like this is what we need to add?
```
const devices = await args.api.get(...
if (devices.length > 0) { 
  request.values.should_be_managed_by__release = devices[0].supervisor_version;
}
```

Reply:
> On a side note, I'm not sure why this isn't just handled by a migration of all existing data that lacks the desired attribute, rather than doing it at the point of PATCHing?

@toochevere this value can't be known a priori due to how flasher images work (first creating a device stub, then populating it with appropriate data during the first boot from non-flasher media).

`supervisor_version` is what the supervisor itself reports back as a string (current state). We then need to look up the corresponding target state and set the value accordingly, so it requires a second lookup for which release is architecture-compatible and matches the version the device has just checked in with. See https://github.com/balena-io/balena-api/blob/ed7ced9ed0b2f4115aab72da6022bb49b2804da8/src/models/hooks/resources/device.ts#L351-L367 for more.